File: modified_binary_search/04_first_and_last_position_of_element_in_sorted_array/solution.js
function search(arr, target, find_first_occurrence = false) {
  if (arr.length === 0) {
    return -1;
  }

  let start = 0;
  let end = arr.length - 1;
  let index_of_interest = -1;
  while (start <= end) {
    let mid = Math.floor((start + end) / 2);
    if (arr[mid] === target) {
      index_of_interest = mid;
      if (find_first_occurrence) {
        end = mid - 1;
      } else {
        start = mid + 1;
      }
    } else if (arr[mid] > target) {
      end = mid - 1;
    } else {
      start = mid + 1;
    }
  }
  return index_of_interest;
}

function searchRange(arr, target) {
  let first_occurrence = search(arr, target, true);
  if (first_occurrence === -1) {
    return [-1, -1];
  }
  let last_occurrence = search(arr, target, false);
  return [first_occurrence, last_occurrence];
}


/* -------------------------------- Example 1 ------------------------------- */
console.log("\n")
console.log("-------------------------------- Example 1 -------------------------------");
let arr = [1, 2, 6, 6, 6, 6, 67, 123, 125]
let target = 6
let expected_output = [2, 5];
let received_output = searchRange(arr, target);
console.log(
  `expected output = ${expected_output.toString()} received output = ${received_output.toString()}`
);
console.assert(expected_output.toString() === received_output.toString(), `incorrect output. expected ${expected_output}, received ${received_output}`)

/* -------------------------------- Example 2 ------------------------------- */
console.log("\n")
console.log("-------------------------------- Example 2 -------------------------------");
arr = [1, 2, 6, 6, 6, 6, 67, 123, 125]
target = 67
expected_output = [6, 6];
received_output = searchRange(arr, target);
console.log(
  `expected output = ${expected_output.toString()} received output = ${received_output.toString()}`
);
console.assert(expected_output.toString() === received_output.toString(), `incorrect output. expected ${expected_output}, received ${received_output}`)

/* -------------------------------- Example 3 ------------------------------- */
console.log("\n")
console.log("-------------------------------- Example 3 -------------------------------");
arr = [1, 2, 6, 6, 6, 6, 67, 123, 125]
target = 195
expected_output = [-1, -1];
received_output = searchRange(arr, target);
console.log(
  `expected output = ${expected_output.toString()} received output = ${received_output.toString()}`
);
console.assert(expected_output.toString() === received_output.toString(), `incorrect output. expected ${expected_output}, received ${received_output}`)

/* -------------------------------- Example 4 ------------------------------- */
console.log("\n")
console.log("-------------------------------- Example 4 -------------------------------");
arr = []
target = 5
expected_output = [-1, -1];
received_output = searchRange(arr, target);
console.log(
  `expected output = ${expected_output.toString()} received output = ${received_output.toString()}`
);
console.assert(expected_output.toString() === received_output.toString(), `incorrect output. expected ${expected_output}, received ${received_output}`)

File: modified_binary_search/04_first_and_last_position_of_element_in_sorted_array/solution.py
def search(arr: list, target: int, find_first_occurence=False):
    if len(arr) == 0:
        return -1

    start = 0
    end = len(arr) - 1
    index_of_interest = -1
    while start <= end:
        mid = (start + end) // 2
        if arr[mid] == target:
            index_of_interest = mid
            if find_first_occurence:
                end = mid - 1
            else:
                start = mid + 1
        elif arr[mid] > target:
            end = mid - 1
        else:
            start = mid + 1
    return index_of_interest


def search_range(arr: list, target: int):
    # Write your code here.
    first_occurence = search(arr, target, find_first_occurence=True)
    if first_occurence == -1:
        return [-1, -1]
    last_occurence = search(arr, target, find_first_occurence=False)
    return [first_occurence, last_occurence]


if __name__ == "__main__":

    # --------------------------------- Example 1 -------------------------------- #
    print("\n")
    print(
        "--------------------------------- Example 1 --------------------------------"
    )
    arr = [1, 2, 6, 6, 6, 6, 67, 123, 125]
    target = 6
    expected_output = [2, 5]
    received_output = search_range(arr, target)
    print(f"Expected Output = {expected_output}, Received Output = {received_output}")
    assert (
        expected_output == received_output
    ), f"Incorrect output. Expected {expected_output}, received {received_output}"

    # --------------------------------- Example 2 -------------------------------- #
    print("\n")
    print(
        "--------------------------------- Example 2 --------------------------------"
    )
    arr = [1, 2, 6, 6, 6, 6, 67, 123, 125]
    target = 67
    expected_output = [6, 6]
    received_output = search_range(arr, target)
    print(f"Expected Output = {expected_output}, Received Output = {received_output}")
    assert (
        expected_output == received_output
    ), f"Incorrect output. Expected {expected_output}, received {received_output}"

    # --------------------------------- Example 3 -------------------------------- #
    print("\n")
    print(
        "--------------------------------- Example 3 --------------------------------"
    )
    arr = [1, 2, 6, 6, 6, 6, 67, 123, 125]
    target = 195
    expected_output = [-1, -1]
    received_output = search_range(arr, target)
    print(f"Expected Output = {expected_output}, Received Output = {received_output}")
    assert (
        expected_output == received_output
    ), f"Incorrect output. Expected {expected_output}, received {received_output}"

    # --------------------------------- Example 4 -------------------------------- #
    print("\n")
    print(
        "--------------------------------- Example 4 --------------------------------"
    )
    arr = []
    target = 5
    expected_output = [-1, -1]
    received_output = search_range(arr, target)
    print(f"Expected Output = {expected_output}, Received Output = {received_output}")
    assert (
        expected_output == received_output
    ), f"Incorrect output. Expected {expected_output}, received {received_output}"
